lib/grpc_provider.cpp

```diff
@@ -0,0 +1,284 @@
+        notifyOperRespondMsg.mutable_cardoperation();
+    cardOperation->set_userid(uid);
+    cardOperation->set_card(card);
+    cardOperation->set_targetuid(targetuid);
+    cardOperation->set_isdrawcard(card == ExplodingKittensProto::NONE ? true
+                                                                      : false);
+    for (auto id : uids) {
+        SendNotifyMsg(roomid, id, notifyOperRespondMsg);
+    }
+}
+
+void Provider::SendSwapResult(int roomid, int uid,
+                              const std::vector<CardType>& cards) {
+    ExplodingKittensProto::NotifyMsg notifySwapRetMsg;
+    ExplodingKittensProto::SwapResult* swapResult =
+        notifySwapRetMsg.mutable_swapresult();
+    *(swapResult->mutable_cards()) =
+        google::protobuf::RepeatedField<google::protobuf::int32>(cards.begin(),
+                                                                 cards.end());
+    SendNotifyMsg(roomid, uid, notifySwapRetMsg);
+}
+
+void Provider::SendExtortResult(int roomid, int uid, CardType card, int srcid,
+                                int dstid) {
+    ExplodingKittensProto::NotifyMsg notifyExtortRetMsg;
+    ExplodingKittensProto::ExtortResult* extortResult =
+        notifyExtortRetMsg.mutable_extortresult();
+    extortResult->set_card(CardTypeToProtoCardType(card));
+    extortResult->set_dstuid(dstid);
+    extortResult->set_srcuid(srcid);
+
+    SendNotifyMsg(roomid, uid, notifyExtortRetMsg);
+}
+
+void Provider::SendDrawResult(int roomid, int uid, CardType card) {
+    ExplodingKittensProto::NotifyMsg notifyDrawRetMsg;
+
+    ExplodingKittensProto::DrawResult* drawResult =
+        notifyDrawRetMsg.mutable_drawresult();
+    drawResult->set_card(CardTypeToProtoCardType(card));
+
+    SendNotifyMsg(roomid, uid, notifyDrawRetMsg);
+}
+
+void Provider::SendPredictResult(int roomid, int uid, int pos) {
+    ExplodingKittensProto::NotifyMsg notifyPredictMsg;
+
+    ExplodingKittensProto::PredictResult* predictResult =
+        notifyPredictMsg.mutable_predictresult();
+    predictResult->set_pos(pos);
+
+    SendNotifyMsg(roomid, uid, notifyPredictMsg);
+}
+
+void Provider::SendSeeThroughResult(int roomid, int uid,
+                                    const std::vector<CardType>& cards) {
+    ExplodingKittensProto::NotifyMsg notifySeeThroughMsg;
+
+    ExplodingKittensProto::SeeThroughResult* seeThroughResult =
+        notifySeeThroughMsg.mutable_seethroughresult();
+    *(seeThroughResult->mutable_cards()) =
+        google::protobuf::RepeatedField<google::protobuf::int32>(cards.begin(),
+                                                                 cards.end());
+    SendNotifyMsg(roomid, uid, notifySeeThroughMsg);
+}
+
+void Provider::SendKO(int roomid, int uid, std::vector<int>& uids) {
+    ExplodingKittensProto::NotifyMsg notifykoMsg;
+
+    ExplodingKittensProto::KO* ko = notifykoMsg.mutable_ko();
+    ko->set_userid(uid);
+    for (auto id : uids) {
```

Comment:
int

Reply:
by the way,  Core is smarter than you imagine.

if you wanna to broadcast a msg.
just use SendNotifyMsg(roomid, 0, notifyKoMsg);

if you send a msg to users except a specific user.
 SendNotifyMsg(roomid, -userId, notifyKoMsg);

Reply:
done